lib/twilioconnection.js

```diff
@@ -436,18 +436,21 @@ class TwilioConnection extends StateMachine {
    * @private
    */
   _handleWelcome({ negotiatedTimeout }) {
-    if (this.state !== 'connecting') {
-      this._log.warn(`Unexpected state "${this.state}" for handling a "welcome"`
+    const log = this._log;
+
+    if (!['connecting', 'waiting'].includes(this.state)) {
+      log.warn(`Unexpected state "${this.state}" for handling a "welcome"`
         + ' message from the TCMP server.');
       return;
     }
 
-    const { maxConsecutiveMissedHeartbeats } = this._options;
+    if (this.state === 'waiting' && this._busyWaitTimeout) {
+      log.debug('Received "welcome" message, no need to retry connection.');
```

Comment:
The check for `this._busyWaitTimeout` is unnecessary because it will be always true when state is waiting.
We never set this._busyWaitTimeout to null. Maybe the check needs to be changed to   `this._busyWaitTimeout.isSet()` or removed altogether.